crates/circuit/src/dag_circuit.rs

```diff
@@ -2803,23 +2808,57 @@ def _format(operand):
     }
 
     /// Returns set of the ancestors of a node as DAGOpNodes and DAGInNodes.
-    fn ancestors(&self, py: Python, node: &DAGNode) -> PyResult<Py<PySet>> {
-        // return {self._multi_graph[x] for x in rx.ancestors(self._multi_graph, node._node_id)}
-        todo!()
+    #[pyo3(name="ancestors")]
+    fn py_ancestors(&self, py: Python, node: &DAGNode) -> PyResult<Py<PySet>> {
+        if let Some(node_index) = node.node {
+            let ancestors = core_ancestors(&self.dag, node_index)
+                .filter_map(|node| self.get_node(py, node).ok());
+            let ancestor_set = PySet::empty_bound(py)?;
+            for ancestor in ancestors {
+                ancestor_set.add(ancestor)?;
+            }
+            return Ok(ancestor_set.unbind());
+        } else {
+            unreachable!("The provided {:?} is not properly initialized.", node)
+        }
     }
 
     /// Returns set of the descendants of a node as DAGOpNodes and DAGOutNodes.
-    fn descendants(&self, py: Python, node: &DAGNode) -> PyResult<Py<PySet>> {
-        // return {self._multi_graph[x] for x in rx.descendants(self._multi_graph, node._node_id)}
-        todo!()
+    #[pyo3(name="descendants")]
+    fn py_descendants(&self, py: Python, node: &DAGNode) -> PyResult<Py<PySet>> {
+        let descendants = self.descendants(node)
+            .filter_map(|node| self.get_node(py, node).ok());
```

Comment:
The `get_node` method will only return an `Err` if something goes wrong when constructing the DAGNode, which really shouldn't happen. But, if it i does, we ought to propagate the error up through this method, rather than turning it into an `Option`.

I imagine the purpose of the `filter_map` you've got here is to ignore any nodes that aren't found. However, if the node ID you provide to `get_node` isn't found, it'll actually _panic_ and crash the process rather than return an `Err`.

Since we're calling `get_node` here using IDs that came from the graph itself, there is no risk of that happening, anyway 🙂.

I'd recommend doing something like:

```rust
let descendants: PyResult<Vec<_>> = self.descendants(node.node.unwrap).map(|node| self.get_node(py, node)).collect();
let descendant_set = PySet::new_bound(py, descendants?)?;
```